src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,448 @@
+{
+    internal sealed class JSProxyContext : IDisposable
+    {
+        private bool _disposedValue;
+
+        // we use this to maintain identity of JSHandle for a JSObject proxy
+        private readonly Dictionary<nint, WeakReference<JSObject>> ThreadCsOwnedObjects = new();
+        // we use this to maintain identity of GCHandle for a managed object
+        private readonly Dictionary<object, nint> ThreadJsOwnedObjects = new(ReferenceEqualityComparer.Instance);
+        // this is similar to GCHandle, but the GCVHandle is allocated on JS side and this keeps the C# proxy alive
+        private readonly Dictionary<nint, PromiseHolder> ThreadJsOwnedHolders = new();
+        // JSVHandle is like JSHandle, but it's not tracked and allocated by the JS side
+        // It's used when we need to create JSHandle-like identity ahead of time, before calling JS.
+        // they have negative values, so that they don't collide with JSHandles.
+        private nint NextJSVHandle = -2;
+        private readonly List<nint> JSVHandleFreeList = new();
+
+#if !FEATURE_WASM_THREADS
+        public static readonly JSProxyContext DefaultInstance = new();
+        public static JSProxyContext CurrentInstance => DefaultInstance;
+        public static JSProxyContext CapturedInstance
+        {
+            get => MainInstance!;
+            set { }
+        }
+        public static JSProxyContext MainInstance => DefaultInstance;
+#else
+        public nint TargetTID;
+        public int ThreadId;
+        public bool IsMainThread;
+        public JSSynchronizationContext SynchronizationContext;
+
+        [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "thread_id")]
+        private static extern ref long GetThreadNativeThreadId(Thread @this);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr GetNativeThreadId()
+        {
+            return (int)GetThreadNativeThreadId(Thread.CurrentThread);
+        }
+
+        // Context of the main thread
+        private static JSProxyContext? _MainInstance;
+        public static JSProxyContext MainInstance
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _MainInstance!;
+            set => _MainInstance = value;
+        }
+
+        // Context captured from parameters of the current operation.
+        // Most of the time, this matches the current thread.
+        // When we pass `JSObject` instance to `JSImport`, it could have thread affinity different than the current thread.
+        // We will use that parameter to capture the target thread to call into.
+        [ThreadStatic]
+        public static JSProxyContext? CapturedInstance;
+
+        // Context of the current thread. Could be null on threads which don't have JS interop, like managed thread pool threads.
+        [ThreadStatic]
+        public static JSProxyContext? CurrentInstance;
+
+        // This is context to dispatch into. In order of preference
```

Comment:
'where to dispatch' seems like it should never be in order of preference. as i understand it, you have three valid dispatch targets:

`anywhere`, for when you are creating a new JS object, or
`anywhere that meets criteria`, for when you are creating a new JS object that requires some condition to be true (i.e. i'm going to fetch so this JS context needs the right cookies and origin)
`a specific thread`, either the thread that owns a given JS object, or the UI thread

`DefaultInstance` feels like it does not encode this information and will just be right most of the time but not always-right-by-construction

Reply:
What rules would you like to see instead ?